File: Python3/801 - 1000/881 Boats to Save People.py
# 156ms, 98.49%; 17.9MB, 14.29%
class Solution:
    def numRescueBoats(self, people: list, limit: int) -> int:
        p = sorted(people)
        i, j, res = 0, len(p) - 1, 0
        while i <= j:
            if p[i] + p[j] <= limit: i += 1
            j -= 1
            res += 1
        return res

File: 801 - 1000/881 Boats to Save People.cc
// 68ms, 100.0%; 14MB, 100.0%
#include <bits/stdc++.h>
using namespace std;

static int x = []() {
    ios_base::sync_with_stdio(false);
    cin.tie(nullptr);
    return 0;
}();

class Solution {
public:
    int numRescueBoats(vector<int> &people, int limit) {
        sort(people.begin(), people.end());
        int i = 0, j = people.size() - 1, res = 0;
        while (i <= j) {
            if (people[i] + people[j] <= limit) ++i;
            --j, ++res;
        }
        return res;
    }
};
